docs/porting-guide.rst

```diff
@@ -2055,6 +2055,28 @@ power down state where as it could be either power down, retention or run state
 for the higher power domain levels depending on the result of state
 coordination. The generic code expects the handler to succeed.
 
+plat\_psci\_ops.pwr\_domain\_suspend\_\early() [optional]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Perform the platform specific actions to prepare to suspend the calling
+CPU and its higher parent power domain levels as indicated by the
+``target_state`` (first argument). It is called by the PSCI ``CPU_SUSPEND``
+API implementation.
+
+The ``target_state`` has a similar meaning as described in
+the ``pwr_domain_off()`` operation. It encodes the platform coordinated
+target local power states for the CPU power domain and its parent
+power domain levels. The handler needs to perform power management operation
+corresponding to the local state at each power level. The generic code
+expects the handler to succeed.
+
+This handler gets called with the data cache enabled to allow platforms
+to perform necessary actions before the CPU actually enters the suspend
+state.
+
+This platform handler is optional and can theoretically replace the
+``pwr_domain_suspend`` handler, when implemented.
+
```

Comment:
For some reason the GitHub UI is still showing the above outdated comment. That should now be ignored.

I don't think the above text properly captures how `pwr_domain_suspend_pwr_down_early()` should be used. Rather than make specific comments, I'm going to suggest alternative text for this section...

This optional function may be used as a performance optimization to replace or complement `pwr_domain_suspend()` on some platforms. Its calling semantics are identical to `pwr_domain_suspend()`, except the PSCI implementation only calls this function when suspending to a power down state, and it guarantees that data caches are enabled.

When `HW_ASSISTED_COHERENCY = 0`, the PSCI implementation disables data caches before calling `pwr_domain_suspend()`. If the `target_state` corresponds to a power down state and it is safe to perform some or all of the platform specific actions in that function with data caches enabled, it may be more efficient to move those actions to this function. When `HW_ASSISTED_COHERENCY = 1`, data caches remain enabled throughout, and so there is no advantage to moving platform specific actions to this function.